### ISSUE TITLE
retainLines option doesn't collapse JSXElements properly

### ISSUE THREAD

**Issue by kevinb7** [babel/babel#1585]
Input code:
```
var React = require("react");
var $_ = require("i18n.js").$_;
var $i18nDoNotTranslate = require("i18n.js").$i18nDoNotTranslate;

function hello_world() {
    return "hello, world!";
}

function goodbye_world() {
    return "goodbye, world";
}

var TestComponent = React.createClass({
    render: function() {
        var maybe = Math.random() > 0.5;

        return <div>
            <$_ textbox={<input type="text"/>}>some text</$_>
            <$i18nDoNotTranslate>other text
                more other text</$i18nDoNotTranslate>
            <$i18nDoNotTranslate>
                {"```if (x < 0) {\n    return;\n}```"}</$i18nDoNotTranslate>
            <$_>before { maybe ? hello_world() : goodbye_world() } after</$_>
            <$_></$_>
        </div>;
    }
});

module.exports = TestComponent;
```

jstransform output:
```
var React = require("react");
var $_ = require("i18n.js").$_;
var $i18nDoNotTranslate = require("i18n.js").$i18nDoNotTranslate;

function hello_world() {
    return "hello, world!";
}

function goodbye_world() {
    return "goodbye, world";
}

var TestComponent = React.createClass({displayName: 'TestComponent',
    render: function() {
        var maybe = Math.random() > 0.5;

        return React.createElement("div", null, 
            React.createElement($_, {textbox: React.createElement("input", {type: "text"})}, "some text"), 
            React.createElement($i18nDoNotTranslate, null, "other text" + ' ' +
                "more other text"), 
            React.createElement($i18nDoNotTranslate, null, 
                "```if (x < 0) {\n    return;\n}```"), 
            React.createElement($_, null, "before ",  maybe ? hello_world() : goodbye_world(), " after"), 
            React.createElement($_, null)
        );
    }
});

module.exports = TestComponent;
```

babel output:
```
var React = require("react");
var $_ = require("i18n.js").$_;
var $i18nDoNotTranslate = require("i18n.js").$i18nDoNotTranslate;

function hello_world() {
    return "hello, world!";}


function goodbye_world() {
    return "goodbye, world";}


var TestComponent = React.createClass({ displayName: "TestComponent", 
    render: function () {
        var maybe = Math.random() > 0.5;

        return React.createElement("div",
            null,
            React.createElement($_,
                { textbox: React.createElement("input", { type: "text" }) },
                "some text"),
            React.createElement($i18nDoNotTranslate,
                null,
                "other text more other text"),
            React.createElement($i18nDoNotTranslate,
                null,
                "```if (x < 0) {\n    return;\n}```"),
            React.createElement($_,
                null,
                "before ",
                maybe ? hello_world() : goodbye_world(),
                " after"),
            React.createElement($_, null));} });module.exports = TestComponent;
```

code used to compile w/ babel:
```
var output_code = babel.transform(code, { 
    retainLines: true, 
    whitelist: whitelist_with_react 
});
```

**Comment by kevinb7** [comment]
I've narrowed down what the issue is.  ```JSXElement.exit(node)``` in build-react-transformer.js is setting ```_prettyCall``` on the call expression node to ```true``` which forces ```CallExpression(node, print)``` in generators/expression.js to adding in new lines.

In order to fix this I've updated ```CallExpression(node,print)``` to look like this:
```
function CallExpression(node, print, parent, format) {
  print(node.callee);

  this.push("(");

  var separator = ",";

  if (node._prettyCall && !format.retainLines) {
    separator += "\n";
    this.newline();
    this.indent();
  } else {
    separator += " ";
  }

  print.list(node.arguments, { separator: separator });

  if (node._prettyCall && !format.retainLines) {
    this.newline();
    this.dedent();
  }

  this.push(")");
}
```

The callsite in ```index.js``` ended up looking like this:
```
this[node.type](node, this.buildPrint(node), parent, this.format);
```

I'll submit a PR for this later today.

**Comment by kevinb7** [comment]
@sebmck Thanks.  I didn't realize that ```this.format``` was accessible there.  It makes sense now that I think about it.  I wrote some tests for this.  Would you like me to submit them?

**Comment by sebmck** [comment]
For sure.

**Comment by sebmck** [comment]
Fixed as of 5.4.5, thanks!
